docs/rfcs/rfc03.md

```diff
@@ -0,0 +1,127 @@
+                    └── 144e62ccbdc9f7d4f8bccc14f33224cbf63e5185
+                        └── k8s.libsonnet
+
+
+### Multiple cache sources
+
+(optional, advanced)
+
+The user can define multiple cache directories.
+They will act as a "search path".
+If a file is not found in the first directory, the next one will be tried until the cache directories are exhausted.
+Only the first cache directory is used to write new files to.
+
+Example:
+
+```bash
+kubecfg --cache-dir ./vendor --cache-dir ~/.cache/kubecfg show ....
+```
+
+Open questions:
+
+* should we always put the default cache at the end of the cache search path?
+
+### Vendoring
+
+To enable vendoring the user needs to pass a custom location for the cache directory and choose a desirable cache pruning policy.
+
+This plays well with the [Flags From Files](https://github.com/kubecfg/kubecfg/pull/224) feature.
+
+```console
+$ cat .kubecfgrc.jsonnet
+{
+    flags+: {
+        'cache-dir': 'vendor',
+        'cache-pruning-policy': 'vendor'
+    },
+}
+$ kubecfg show foo.jsonnet
+```
+
+### Cache pruning
+
+The simplest thing to do is to let the cache just grow and grow and let the user just clean it up when needed.
+
+The other strategies are:
+
+* size limit, delete least recently used
+* remove all unreferenced
+* ???
+
+#### Remove all unreferenced strategy
+
+One difficulty of the "remove all unreferenced" mechanism is that some repos contain multiple top level jsonnet files,
+and if the cache pruning was executed after every invocation of the `kubecfg` command then the only cached/vendored files left would be 
+the ones pertaining the last evaluated file.
+
+I see a few options for this:
+
+1. an explicit "prune" step where you tell kubecfg to look at all the jsonnet files and run the find-deps on them and delete all cache files that aren't visited.
+
+2. at the beginning for the batch execution of multiple kubecfg commands, create a new (physical or logica) cache/vendor directory and filling this new cache by doing the same operation that you'd do if you were filling the cache after cache misses (except instead of downloading the files from remote sources we'd read them from the previous cache location that would be passed to the cache search path).
+
+3. every kubecfg invocation of kubecfg will be passed a `--cache-access-log` flag which will reference a tmp file to which kubecfg will append the filenames of the cache files it has touched. At the end of the multi-file generation script, that file will be used to compute which cache files can be deleted (e.g. by callign a command like `kubecfg cache prune --not-in "${cache_access_log}"`)
+
+### CI
+
+When running in a CI, it's advisable to rely solely on the vendored files and break the build if the user
+accidentally forgot to commit new imports:
+
+```bash
+kubecfg --ci ....
+```
+
+The `--ci` flag is intentionally generic because it may govern other behaviour that is sensible in a CI environment.
+
+### Filename escaping
```

Comment:
Urgh.  I like making the URLs "understandable", particularly if we're asking folks to rm -r files from the cache directly.

An alternative (that I don't like, but mentioning for completeness) is that we just base64 or sha hash the URL (including scheme, path separators, etc), and the original URLs become opaque filenames at the filesystem level.

Reply:
I think I brainfarted above:

All we have to do is to decouple foundAt from the cacheKey.